src/server/requestHandlers.js

```diff
@@ -1,431 +1,490 @@
+    gradeNames: "kettle.request.http",
+    invokers: {
+        handleRequest: {
+            funcName: "sjrk.storyTelling.server.handleGetStory",
+            args: ["{request}", "{server}.storyDataSource", "{server}.options.secureConfig.uploadedFilesHandlerPath"]
+        }
+    }
+});
+
+sjrk.storyTelling.server.handleGetStory = function (request, dataSource, uploadedFilesHandlerPath) {
+    var id = request.req.params.id;
+    var promise = dataSource.get({directStoryId: id});
+
+    promise.then(function (response) {
+
+        fluid.transform(response.content, function (block) {
+            if (block.blockType === "image") {
+                if (block.imageUrl) {
+                    block.imageUrl = uploadedFilesHandlerPath + "/" + block.imageUrl;
+                }
+                return block;
+            } else if (block.blockType === "audio" || block.blockType === "video") {
+                if (block.mediaUrl) {
+                    block.mediaUrl = uploadedFilesHandlerPath + "/" + block.mediaUrl;
+                }
+                return block;
+            }
+        });
+
+        var responseAsJSON = JSON.stringify(response);
+        request.events.onSuccess.fire(responseAsJSON);
+    }, function (error) {
+        var errorAsJSON = JSON.stringify(error);
+        request.events.onError.fire({
+            isError: true,
+            message: errorAsJSON
+        });
+    });
+};
+
+fluid.defaults("sjrk.storyTelling.server.saveStoryWithBinariesHandler", {
+    gradeNames: "kettle.request.http",
+    requestMiddleware: {
+        "saveStoryWithBinaries": {
+            middleware: "{server}.saveStoryWithBinaries"
+        }
+    },
+    invokers: {
+        handleRequest: {
+            funcName: "sjrk.storyTelling.server.handleSaveStoryWithBinaries",
+            args: ["{arguments}.0", "{server}.storyDataSource", "{server}.options.globalConfig.authoringEnabled"]
+        }
+    }
+});
+
+sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSource, authoringEnabled) {
+    if (authoringEnabled) {
+        var rotateImagePromises = [];
+
+        // rotate any images based on their EXIF data, if present
+        fluid.each(request.req.files.file, function (singleFile) {
+            if (singleFile.mimetype && singleFile.mimetype.indexOf("image") === 0) {
+                var singleRotatePromise = sjrk.storyTelling.server.rotateImageFromExif(singleFile).then(null, function (error) {
+                    fluid.log(fluid.logLevel.WARN, "Image rotation failed for file " + singleFile.path + ": " + error.message);
+                });
+
+                rotateImagePromises.push(singleRotatePromise);
+            }
+        });
+
+        // using [""] notation for .catch() and .finally() to pass linting,
+        // at least until FLUID-6419 is addressed:
+        // https://issues.fluidproject.org/browse/FLUID-6419
+        Promise.all(rotateImagePromises)["catch"](function (error) {

```

Comment:
`rotateImagePromises` contains an array of infusion promises. Is there a reason that you decided to use JS standard `Promise.all()` instead of `[infusion fluid.promise.sequence()](https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisesequencesources-options)` to handle it? Using fluid.promise.sequence(), functions attached for `catch` and `finally` can be more proper `resolve` and `reject` callback functions.

Reply:
The idea was to wait until all the promises had finished resolving, but that the final state of them is not relevant, since some of the promises will reject with an "error" message even if everything is okay. E.g. if the image provided has no EXIF data and nothing needs to be done, or if the image is already at Orientation 1.

I suppose I could use `fluid.promise.sequence()` and then chain the `finally` and `catch` off that last one, but do `fluid` promises have those methods? I thought that wasn't the case.

Reply:
Best to fix up the promise source so that it doesn't reject when everything is fine.

Reply:
I agree with @amb26's suggestion, though complication is introduced by the fact that my call to `jo.rotate` within `rotateImageFromExif()` will return an error code and reject its promise even if the result is acceptable. Here's the documentation, in case you're curious about all the possible cases: https://github.com/johansatge/jpeg-autorotate#error-handling

With that in mind, would it be reasonable to have a check on the error codes returned and, in the specific cases where everything is okay, resolve the promise and keep on rolling?
i.e. at this point: https://github.com/BlueSlug/sjrk-story-telling/blob/b084f7180f84cf406b82e518134711dfe5702669/src/server/requestHandlers.js#L232

If I add this error-code checking and resolve in those specific cases that are acceptable, then I could use `fluid.promise.sequence` and remove the awkward syntax workaround and ES6 class.

Reply:
Yes, I think that is the best approach